File: MAP(Advanced Programming Methods)/LabAssig7/LabAssig7/Model/Statements/CloseFileStatement.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using LabAssig7.Model;
using LabAssig7.Model.ADT;
using LabAssig7.Model.Expressions;
using System.IO;
namespace LabAssig7.Model.Statements
{
    public class CloseFileStatement:Statement
    {
        private Expression expr;

        public CloseFileStatement(Expression e)
        {
            this.expr = e;
        }

        public PrgState Execute(PrgState p)
        {
            int res = this.expr.Eval(p.Dict);
            if (!p.FileTable.Contains(res))
                throw new SystemException("File not found");
            FileData fd = p.FileTable.GetValue(res);
            StreamReader header = fd.Header;
            header.Close();
            p.FileTable.Remove(res);
            return p;
        }

        public override string ToString()
        {
            return "CloseFile(" + this.expr + ")\n";
        }
    }
}

File: MAP(Advanced Programming Methods)/lab4/src/Model/Statement/CloseFileStatement.java
package Model.Statement;

import Exceptions.FileException;
import Model.ADT.IFileTable;
import Model.Expression.Expression;
import Model.FileData;
import Model.PrgState;

import java.io.BufferedReader;
import java.io.IOException;

public class CloseFileStatement implements Statement {

    private Expression fileDescriptor;

    public CloseFileStatement(Expression fd)
    {
        this.fileDescriptor =fd;
    }

    @Override
    public PrgState execute(PrgState p)
    {
        try
        {
            int descriptor = this.fileDescriptor.eval(p.getSymbolTable(),p.getHeap());
            IFileTable<Integer,FileData> file =p.getFileTable();
            BufferedReader buff =   file.get(descriptor).getHeader();
            buff.close();
            file.remove(descriptor);
            return p;
        }
        catch (IOException e)
        {
            throw new FileException(e.toString()+"Eroare close file");
        }

    }

    @Override
    public String toString() {
        return "CloseFile("+this.fileDescriptor+")";
    }
}

